File: Exercise/LeetCode/Array/P367.java
package Array;

public class P367 {
    public boolean isPerfectSquare(int num) {
        int middle = 1;
        while (middle * middle < num) {
            if (middle * 2 < 0) {
                break;
            }
            middle *= 2;
        }
        for (int i = middle / 2; i <= middle; i++) {
            if (i * i == num) {
                return true;
            }
            if (i * i > num) {
                return false;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        System.out.println(new P367().isPerfectSquare(2000105819));
    }
}

File: Exercise/LeetCode/Array/P367.cpp
#include <iostream>
#include <algorithm>
#include <math.h>
#include <vector>
#include <string>

using namespace std;
class Solution
{
public:
    bool isPerfectSquare(int num)
    {
        long middle = 1;

        while (middle * middle < num)
        {
            if (middle * 2 < 0)
            {
                break;
            }
            middle *= 2;
            if (middle * middle == num)
            {
                return true;
            }
        }

        for (long i = middle / 2; i <= middle; i++)
        {
            if (i * i == num)
            {
                return true;
            }
            if (i * i > num)
            {
                return false;
            }
        }

        return false;
    }
};
int main(void)
{
    cout << Solution().isPerfectSquare(16) << endl;
    return 0;
}